### PR TITLE
perf: analyze cjs exports and emit typescript in parallel

### PR DESCRIPTION
**Pull request by dsherret**
Two main things we can do in parallel:

1. CJS export analysis
1. TypeScript emit

This makes the loader async to make doing this easier.

Deleting the node analysis cache between runs and importing some npm packages (this only makes the first run much faster due to the caches... the second run is the same speed as befroe):

```
Benchmark 1: ../deno/deno_old run -A ts_morph.ts
  Time (mean ± σ):     406.1 ms ±   6.5 ms    [User: 358.1 ms, System: 62.2 ms]
  Range (min … max):   395.9 ms … 412.9 ms    10 runs
 
Benchmark 2: ../deno/target/release/deno run -A ts_morph.ts
  Time (mean ± σ):     278.2 ms ±   9.3 ms    [User: 411.1 ms, System: 81.4 ms]
  Range (min … max):   266.1 ms … 292.9 ms    10 runs
 
Summary
  ../deno/target/release/deno run -A ts_morph.ts ran
    1.46 ± 0.05 times faster than ../deno/deno_old run -A ts_morph.ts
% cat ts_morph.ts
import "npm:ts-morph@22.0.0";
import "npm:ts-morph@21.0.0";
import "npm:ts-morph@20.0.0";
import "npm:ts-morph@19.0.0";
```

Deleting the TypeScript emit and module analysis cache between runs and importing multiple copies of Dax:

```
Benchmark 1: ../deno/deno_old run -A dax_deno_land_x.ts
  Time (mean ± σ):     587.3 ms ±  24.0 ms    [User: 347.5 ms, System: 221.0 ms]
  Range (min … max):   569.2 ms … 650.7 ms    10 runs
 
Benchmark 2: ../deno/target/release/deno run -A dax_deno_land_x.ts
  Time (mean ± σ):     489.3 ms ±  14.3 ms    [User: 554.8 ms, System: 411.2 ms]
  Range (min … max):   473.9 ms … 510.2 ms    10 runs
 
Summary
  ../deno/target/release/deno run -A dax_deno_land_x.ts ran
    1.20 ± 0.06 times faster than ../deno/deno_old run -A dax_deno_land_x.ts
scratch % cat dax_deno_land_x.ts
import "https://deno.land/x/dax@0.39.2/mod.ts";
import "https://deno.land/x/dax@0.38.0/mod.ts";
import "https://deno.land/x/dax@0.37.0/mod.ts";
import "https://deno.land/x/dax@0.36.0/mod.ts";
import "https://deno.land/x/dax@0.35.0/mod.ts";
```

Note that this is still parsing typescript files sequentially because it doesn't include (https://github.com/denoland/deno_graph/pull/477). It is just emitting typescript files in parallel.

### CODE REVIEW

File: ext/node/analyze.rs [hunk 1]
```diff
@@ -113,6 +117,7 @@ impl<TCjsCodeAnalyzer: CjsCodeAnalyzer> NodeCodeTranslator<TCjsCodeAnalyzer> {
       reexports_to_handle.push_back((reexport, specifier.clone()));
     }
 
+    // todo(dsherret): we could run this analysis concurrently in a FuturesOrdered
```
Comment (dsherret) [inline ext/node/analyze.rs:120]: This is a future perf improvement.

File: cli/emit.rs [hunk 2]
```diff
@@ -39,16 +39,16 @@ impl Emitter {
...
       transpile_and_emit_options_hash,
     }
   }
 
-  pub fn cache_module_emits(
+  pub async fn cache_module_emits(
     &self,
     graph: &ModuleGraph,
   ) -> Result<(), AnyError> {
+    // todo(dsherret): we could do this concurrently
```
Comment (dsherret) [inline cli/emit.rs:51]: Another future perf improvement.

File: cli/module_loader.rs [hunk 3]
```diff
@@ -574,27 +577,98 @@ impl<TGraphContainer: ModuleGraphContainer> CliModuleLoader<TGraphContainer> {
...
+          code: transpile_result,
+          found_url: specifier.clone(),
+          media_type,
+        })
+      }
+      Err(err) => Err(err),
+    }
+  }
+
+  fn load_prepared_module_sync(
```
Comment (bartlomieju) [inline cli/module_loader.rs:616]: Is this temporary or will we always need both sync and async versions?

File: cli/module_loader.rs [hunk 4]
```diff
@@ -685,15 +774,22 @@ impl<TGraphContainer: ModuleGraphContainer> ModuleLoader
...
-      requested_module_type,
-    ))
+    let inner = self.0.clone();
+    let specifier = specifier.clone();
+    let maybe_referrer = maybe_referrer.cloned();
+    deno_core::ModuleLoadResponse::Async(
+      async move {
+        inner
+          .load_inner(
+            &specifier,
```
Comment (bartlomieju) [inline cli/module_loader.rs:784]: Very cool 👍

File: tests/integration/inspector_tests.rs
Comment (bartlomieju) [inline tests/integration/inspector_tests.rs:1]: Drive-by fix?

**Review by bartlomieju — APPROVED** [review]
LGTM

File: tests/integration/inspector_tests.rs
Comment (dsherret) [reply tests/integration/inspector_tests.rs:1]: Since it's async these now happen out of order sometimes.

File: cli/module_loader.rs (hunk 3 above)
Comment (dsherret) [reply cli/module_loader.rs:616]: Source map lookup is sync, so we'll probably need this unless we move the spawn blocking to a higher level (I investigated doing this just now, and the performance is equivalent, but it means that all our code in the worker needs to be Send/Sync, which is something I don't think we want to do)
